### PR TITLE
Weapon refactor

### PR DESCRIPTION
**Pull request by JohnoRobson**
Removed the default weapon and added a new one called "SMG." Actors no longer spawn with weapons, instead they must be picked up by walking into them. Weapons are dropped on death. Ammo is not implemented. A new AI state was added for finding a weapon when an AI actor does not have one. Fixed a bug where weapon knockback only applied if the actor wasn't trying to move.

The Weapon class extends the new GameItem class, which is intended to be used for things that exist in the game world that actors can pick up or that can be stored in an inventory, such as weapons, health, and ammo.

_Unrelated bug fix:_ The FindHealthState was bugged so that it would never exit once it was entered. Now there's a max health check that sets the state back to FindEnemy if the actor has max health.

### CODE REVIEW

File: scripts/world.gd [hunk 1]
```diff
@@ -79,6 +92,15 @@ func get_closest_available_health(from_position: Vector3) -> HealthPickup:
...
+	pickups = pickups.filter(func(a): return a.health_is_available)
 	return pickups.front()
+
+func get_closest_available_weapon(from_position: Vector3) -> Weapon:
+	var weapon_array = []
+	weapon_array.append_array(weapons)
+
+	weapon_array.sort_custom(func(a, b): return from_position.distance_to(a.global_transform.origin) < from_position.distance_to(b.global_transform.origin))
+	weapon_array = weapon_array.filter(func(a): return !a.is_held)
+	return weapon_array.front()
```
Comment (morganwynne) [inline scripts/world.gd:106]: The front() array method will throw an error if the array is empty. Let's first check if the array contains any values and return null if not.
```
E 0:00:07:0268   world.gd:106 @ get_closest_available_weapon(): Can't take value from empty array.
  <C++ Error>    Condition "_p->array.size() == 0" is true. Returning: Variant()
  <C++ Source>   core/variant/array.cpp:319 @ front()
  <Stack Trace>  world.gd:106 @ get_closest_available_weapon()
                 find_weapon_state.gd:11 @ execute()
                 state_machine.gd:19 @ _process()
                 ai_actor_controller.gd:14 @ _process()
```

**Review by morganwynne — CHANGES_REQUESTED** [review]

**Comment by morganwynne** [comment]
Let's start organizing Pull Requests by keeping the PR assigned to the person responsible for the code and adding the other as a Reviewer so we get a chance to review it before it gets merged into main. Does that sound good?

**Comment by JohnoRobson** [comment]
> Let's start organizing Pull Requests by keeping the PR assigned to the person responsible for the code and adding the other as a Reviewer so we get a chance to review it before it gets merged into main. Does that sound good?
> 
> On that note, I am very much still learning how GitHub works, but I do like it a lot.

Sounds good to me!

File: scripts/world.gd (hunk 1 above)
Comment (JohnoRobson) [reply scripts/world.gd]: I've added ternary checks for this and the other two object finding methods in World.

**Review by morganwynne — APPROVED** [review]

**Pull request merged**